sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/TDigestQuantiles.java

```diff
@@ -0,0 +1,345 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sketching;
+
+import com.google.auto.value.AutoValue;
+import com.tdunning.math.stats.MergingDigest;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.CoderRegistry;
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.transforms.Combine;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+
+
+/**
+ * {@code PTransform}s for getting information about quantiles in a stream.
+ *
+ * <p>This class uses the T-Digest structure introduced by Ted Dunning, and more precisely
+ * the {@link MergingDigest} implementation.
+ *
+ * <h2>References</h2>
+ *
+ * <p>The paper and implementation are available on Ted Dunning's
+ * <a href="https://github.com/tdunning/t-digest">Github profile</a>
+ *
+ * <h2>Parameters</h2>
+ *
+ * <p>Only one parameter can be tuned in order to control the tradeoff between
+ * the estimation accuracy and the memory use. <br>
+ *
+ * <p>Stream elements are compressed into a linked list of centroids.
+ * The compression factor {@code cf} is used to limit the number of elements represented by
+ * each centroid as well as the total number of centroids. <br>
+ * The relative error will always be a small fraction of 1% for values at extreme quantiles
+ * and in general {@code <1000 ppm} at middle quantiles. <br>
+ *
+ * <p>By default the compression factor is set to 100. In the paper, experiments suggest that
+ * values at mid quantiles will have an accuracy of 1% for streams up to 100k elements. <br>
+ * If you have a larger stream or want a better accuracy,
+ * you should increase the compression factor. <br>
+ * For instance a value of 500 would guarantee a relative error of less than 1% in streams
+ * up to 10M elements. Note that a value of 1000 is extremely large.
```

Comment:
Hmm not sure I understand: what if this is a streaming pipeline with an infinite number of elements over time: is the accuracy going to decrease over time? Is a value of 1000 enough for all practical purposes?

If the default value of 100 is only good for up to 100k elements, that seems too small for typical use cases of Beam: with 100k elements you can trivially compute exact quantiles and you don't need an approximate quantile algorithm. I'd recommend to set the default to a value that's enough for all practical purposes, and say something like "if you want less memory usage, set it lower; e.g. ..."

Reply:
As stated in the papers, the relative error seems to not depend on the size of the stream. 
IMO the compression factor should be chose depending on the accuracy we want to guarantee on mid quantiles such as the median, because the sketch will always be extremely accurate on extreme quantiles.
That's why one would want to increase the factor to 200 or 500, to guarantee for instance a relative error of less than 1% for the median.
In all cases as I said in a comment below we can certainly use the "3/c" formula to define the relative error, because it is used in Streamlib's tdigest based on QDistest and this implementation should always be more accurate than these ones.

Reply:
I am confused: you say that the relative error does not depend on size of the stream, and yet the javadoc comment says that accuracy of 1% requires compression factor 100 in streams up to 100k elements but 500 in streams up to 10M elements. Am I missing something?

Reply:
I am going to change the javadoc in the next commit.
What does not depend on the stream size is the max error bound of 3/c. But the real accuracy will always be less than this value and the real value will depend on the compression factor. If we take the example I give in the javadoc, a compression factor of 100 gives a max error bound of 3%, and yet the real accuracy is 1%. What I say is that we can still improve this accuracy by increasing the compression factor, but I should not talk about the stream size.